### PR TITLE
Use FindCUDAToolkit to find cuda dependencies

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83199
* #76104
* __->__ #82695

### CODE REVIEW

**Comment by peterbell10** [comment]
@malfet ping

**Review by malfet — APPROVED** [review]

**Review by malfet — APPROVED** [review]
Looks good to me (GH was glitching)

**Comment by malfet** [comment]
@pytorchbot revert -m "Break cublas packaging into wheel" -c nosignal

After the change `python -c "import torch;x=torch.rand(3, 3, device='cuda');print(torch.linalg.svd(torch.mm(x, x.t())))"` started to fail with 
```
RuntimeError: Error in dlopen: /home/nshulga/miniconda3/envs/test-py37/lib/python3.7/site-packages/torch/lib/libtorch_cuda_linalg.so: undefined symbol: cudsh_slascl_
```

Checked by downloading build artifacts for https://github.com/pytorch/pytorch/actions/runs/3198549830 and comparing it with one from parent commit https://github.com/pytorch/pytorch/actions/runs/3198518074

**Comment by malfet** [comment]
Also, this change increased wheel size from 1.8G to 2.12G
cc: @ZainRizvi , @seemethere we had some sort of mechanism for detecting binary size spikes, don't we?

**Comment by seemethere** [comment]
> Also, this change increased wheel size from 1.8G to 2.12G cc: @ZainRizvi , @seemethere we had some sort of mechanism for detecting binary size spikes, don't we?

I believe we do, let me check if it's still working.

It's not :/ [onedetection link](https://www.internalfb.com/onedetection/alerts?observer_id=338021534200568)

File: cmake/public/cuda.cmake [hunk 1]
```diff
@@ -226,84 +243,44 @@ endif()
...
-    # Add explicit dependency to cudart_static to fix
-    # libcublasLt_static.a.o): undefined reference to symbol 'cudaStreamWaitEvent'
-    # error adding symbols: DSO missing from command line
+        # NOTE: cublas is always linked dynamically
+        CUDA::cublas CUDA::cublasLt)
     set_property(
-      TARGET caffe2::cublas APPEND PROPERTY INTERFACE_LINK_LIBRARIES
-      "${CUDA_cudart_static_LIBRARY}" rt dl)
+        TARGET caffe2::cublas APPEND PROPERTY INTERFACE_LINK_LIBRARIES
+        CUDA::cudart_static rt)
```
Comment (peterbell10) [inline cmake/public/cuda.cmake:278]: This is the cause of the binary size difference. The existing code links cublas dynamically, even in the `CAFFE2_STATIC_LINK_CUDA` path.

File: caffe2/CMakeLists.txt [hunk 2]
```diff
@@ -978,18 +979,18 @@ elseif(USE_CUDA)
     )
     if($ENV{ATEN_STATIC_CUDA})
       target_link_libraries(torch_cuda_linalg PRIVATE
-          ${CUDA_TOOLKIT_ROOT_DIR}/lib64/libcusolver_static.a
-          ${CUDA_TOOLKIT_ROOT_DIR}/lib64/liblapack_static.a     # needed for libcusolver_static
+          CUDA::cusolver_static
+          ${CUDAToolkit_LIBRARY_DIR}/liblapack_static.a     # needed for libcusolver_static
```
Comment (peterbell10) [inline caffe2/CMakeLists.txt:983]: And this library was missing, causing svd to fail when cusolver was linked statically.

**Comment by peterbell10** [comment]
@malfet ping

**Comment by BowenBao** [comment]
Getting `Failed to find nvToolsExt` error on windows with cuda 11.8 after this commit. Maybe it has to do with the newer cuda version I'm using. Builds successfully after reverting.

**Comment by peterbell10** [comment]
@BowenBao where is nvToolsExt installed for you?

**Comment by BowenBao** [comment]
@peterbell10 
```
> dir env: | grep NVTOOLS
NVTOOLSEXT_PATH                C:\Program Files\NVIDIA Corporation\NvToolsExt\
```

**Comment by peterbell10** [comment]
Hrm that should be one of the default search paths on windows , and we check the `NVTOOLSEXT_PATH` environment variable as well.

@BowdenBao can you try deleting the build folder, running cmake with `CMAKE_FIND_DEBUG_MODE=ON`, and share the log?

File: cmake/Caffe2Config.cmake.in [hunk 3]
```diff
@@ -85,7 +85,13 @@ if(@USE_CUDA@)
   # be found again when including the Caffe2 target.
   set(CAFFE2_USE_CUDA @USE_CUDA@)
   set(CAFFE2_USE_TENSORRT @USE_TENSORRT@)
+
+  # Add current directory to module path so we pick up FindCUDAToolkit.cmake
+  set(old_CMAKE_MODULE_PATH CMAKE_MODULE_PATH)
+  list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}")
   include("${CMAKE_CURRENT_LIST_DIR}/public/cuda.cmake")
+  set(CMAKE_MODULE_PATH old_CMAKE_MODULE_PATH)
```
Comment (mthrok) [inline cmake/Caffe2Config.cmake.in:93]: @peterbell10 This is assigning the string 'old_CMAKE_MODULE_PATH' to the 'CMAKE_MODULE_PATH'.
This is causing issues in downstream projects.

https://github.com/pytorch/audio/actions/runs/4287763310/jobs/7469002979#step:8:529

**Comment by BowenBao** [comment]
@peterbell10 the issue seems to be around 

https://github.com/pytorch/pytorch/blob/961e05245c47219800fd55b8da035ae7dcbe564b/cmake/Modules/FindCUDAToolkit.cmake#L1045-L1055. 

The log where I set `NVTOOLSEXT_PATH` in my environment
```
CMake Debug Log at cmake/Modules/FindCUDAToolkit.cmake:1050 (find_path):
  find_path called with the following settings:

    VAR: CUDAToolkit_nvToolsExt_INCLUDE_DIR
    NAMES: "nvToolsExt.h"
    ...

  find_path considered the following locations:

    C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.8/include/include/nvToolsExt.h
    C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.8/include/nvToolsExt.h
    C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.8/include/nvToolsExt.h
    C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.8/nvToolsExt.h
    C:/Program Files/NVIDIA Corporation/NvToolsExt;C:/Program Files/NVIDIA Corporation/NvToolsExt/include/nvToolsExt.h
    C:/Program Files/NVIDIA Corporation/NvToolsExt;C:/Program Files/NVIDIA Corporation/NvToolsExt/nvToolsExt.h

  The item was not found.
```

Removing `NVTOOLSEXT_PATH` and re-run

```
CMake Debug Log at cmake/Modules/FindCUDAToolkit.cmake:1050 (find_path):
  find_path called with the following settings:

    VAR: CUDAToolkit_nvToolsExt_INCLUDE_DIR
    NAMES: "nvToolsExt.h"
    ...

  find_path considered the following locations:

    C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.8/include/include/nvToolsExt.h
    C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.8/include/nvToolsExt.h
    C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.8/include/nvToolsExt.h
    C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.8/nvToolsExt.h

  The item was found at

    C:/Program Files/NVIDIA Corporation/NvToolsExt/include/nvToolsExt.h
```

**Comment by peterbell10** [comment]
@BowenBao thanks for your help. Would you mind testing the latest commit on this PR?

**Comment by BowenBao** [comment]
@peterbell10 worked like a charm, thanks!

File: cmake/public/cuda.cmake [hunk 4]
```diff
@@ -39,15 +39,36 @@ endif()
...
+if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.12.0)
+  cmake_policy(SET CMP0074 NEW)
+endif()
+
+find_package(CUDAToolkit REQUIRED)
+
+cmake_policy(POP)
+
+if(NOT CMAKE_CUDA_COMPILER_VERSION STREQUAL CUDAToolkit_VERSION OR
+    NOT CUDA_INCLUDE_DIRS STREQUAL CUDAToolkit_INCLUDE_DIR)
```
Comment (ax3l) [inline cmake/public/cuda.cmake:62]: I posted a fix for a regression from these lines in #108932

**Pull request closed without merging**